Spigot-Server-Patches/0622-Strip-raytracing-for-EntityLiving-hasLineOfSight.patch

```diff
@@ -0,0 +1,136 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Paul Sauve <paul@technove.co>
+Date: Sat, 31 Oct 2020 18:43:02 -0500
+Subject: [PATCH] Strip raytracing for EntityLiving#hasLineOfSight
+
+The IBlockAccess#rayTrace method is very wasteful in both allocations,
+and in logic. While EntityLiving#hasLineOfSight provides static
+parameters for collisions with blocks and fluids, the method still does
+a lot of dynamic checks for both of these, which result in extra work.
+As well, since the fluid collision option is set to NONE, the entire
+fluid collision system is completely unneeded, yet used anyways.
+
+This patch was originally created for the AirplaneLite project (https://github.com/Technove/AirplaneLite) under the GPLv3 License.
+
+diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
+index 0294ff10dc8edb24aabfbe1589048a405c356ef0..e65f802f376f3c9debbd0ccc76e57a411c777405 100644
+--- a/src/main/java/net/minecraft/server/EntityLiving.java
++++ b/src/main/java/net/minecraft/server/EntityLiving.java
+@@ -2979,7 +2979,7 @@ public abstract class EntityLiving extends Entity {
+         Vec3D vec3d = new Vec3D(this.locX(), this.getHeadY(), this.locZ());
+         Vec3D vec3d1 = new Vec3D(entity.locX(), entity.getHeadY(), entity.locZ());
+ 
+-        return this.world.rayTrace(new RayTrace(vec3d, vec3d1, RayTrace.BlockCollisionOption.COLLIDER, RayTrace.FluidCollisionOption.NONE, this)).getType() == MovingObjectPosition.EnumMovingObjectType.MISS;
++        return this.world.rayTraceDirect(vec3d, vec3d1, VoxelShapeCollision.a(this)) == MovingObjectPosition.EnumMovingObjectType.MISS; // Paper - use direct method
```

Comment:
VoxelShapeCollieion#a Needs OBFHelper

Reply:
The reason I didn't use obfhelpers here is because the resulting code isn't hand written or needing to be maintained across versions. When the version updates, you'll ideally have to strip it down again (which is fairly simple itself.)

Reply:
The way I understand the general rule for OBFHELPERs is if you include a new method in a line that wasn't there before (in this case VoxelShapeCollision#a) you should add an OBFHELPER for is.

Reply:
This would indeed need an obf helper.